### DISCUSSION TITLE
Cannot restore ODF correctly from file exported by mtex

### DISCUSSION THREAD

**Discussion by YoshieraHuang** [mtex-toolbox/mtex#493] [Ask Anything] [answered]
I use mtex to export a simple ODF to file and try to restore the ODF from the file. I have two version of mtex in my hand: `5.1.1` and 
 `5.3.1`. `5.1.1` can restore the ODF correctly but `5.3.1` cannot. Is this a bug or i use mtex incorrectly?

Here are the details:

I use this script to export ODF:
```matlab
fname = '/home/jasper/Desktop/Cu.recrystal.odf.txt';
CS = loadCIF('Cu-Copper.cif');
SS = specimenSymmetry('1');
mod1 = orientation.cube(CS, SS);
odf = unimodalODF(mod1, 'halfwidth', 5*degree);
textureindex(odf)
h = [Miller(1,0,0,CS),Miller(1,1,0,CS),Miller(1,1,1,CS)];
% 
% % Pole figure
w1 = figure;
plotPDF(odf,h,'minmax')
mtexColorbar       % create a new colorbar\
set(gcf,'PaperPositionMode','auto');
ofname = [fname,'.PF.png']
print(w1,'-dpng',ofname);
 

% Inverse pole figure
w2 = figure;
plotIPDF(odf,[xvector,yvector,zvector],'minmax')
mtexColorbar       % create a new colorbar\
set(gcf,'PaperPositionMode','auto');
ofname = [fname,'.IPF.png'];
print(w2,'-dpng',ofname);

% define a equispaced grid in orientation space with resolution of 5 degree
S3G = equispacedSO3Grid(CS,'resolution',5*degree);
export(odf, fname, S3G, 'Bunge', 'generic');
%end
``` 

Here are the PFs and IPFs generated from the script above:
![Cu recrystal odf 5 1 1 txt g PF](https://user-images.githubusercontent.com/38752027/85922395-759c7280-b8b5-11ea-9035-20ed76b0184e.png)
![Cu recrystal odf txt g IPF](https://user-images.githubusercontent.com/38752027/85922398-7a612680-b8b5-11ea-8ab9-b0cb65c57060.png)

Here is my script to restore ODF:
```matlab
%Import script
clear all;
fname = '/home/jasper/Desktop/Cu.recrystal.odf.txt';

CS = loadCIF('Cu-Copper.cif');
odf = loadODF_generic(fname, 'cs', CS, ...
    'ColumnNames', {'phi1', 'Phi', 'phi2', 'weights'});
textureindex(odf)
h = [Miller(1,0,0,CS),Miller(1,1,0,CS),Miller(1,1,1,CS)];

% % Pole figure
w1 = figure;
plotPDF(odf,h,'minmax')
mtexColorbar       % create a new colorbar\
set(gcf,'PaperPositionMode','auto');
ofname = [fname,'.PF.png']
print(w1,'-dpng',ofname);
 

% Inverse pole figure
w2 = figure;
plotIPDF(odf,[xvector,yvector,zvector],'minmax')
mtexColorbar       % create a new colorbar\
set(gcf,'PaperPositionMode','auto');
ofname = [fname,'.IPF.png'];
print(w2,'-dpng',ofname);
```

in 5.1.1, the generated PFs and IPFs is consistent:
![Cu recrystal odf 5 1 1 txt PF](https://user-images.githubusercontent.com/38752027/85922415-995fb880-b8b5-11ea-8a44-db39ebcbc3c6.png)
![Cu recrystal odf 5 1 1 txt IPF](https://user-images.githubusercontent.com/38752027/85922417-9c5aa900-b8b5-11ea-97b4-b019d5aeb975.png)

However, in 5.3.1, the generated PFs and IPFs is weird:
![Cu recrystal odf txt5 3 1 PF](https://user-images.githubusercontent.com/38752027/85938618-8db9d380-b941-11ea-94a6-a374e0dbebc3.png)
![Cu recrystal odf txt5 3 1 IPF](https://user-images.githubusercontent.com/38752027/85938619-8eeb0080-b941-11ea-8a16-b6ebcdd9d774.png)

the orientation densities(OD) of three equivalent directions <100>, <010> and <001> is not equivalent now.

**Comment by ralfHielscher** [comment]
Adding the option `'lsqnonneg'` should resolve your issue, i.e.,
``` matlab
odf = loadODF_generic(fname, 'cs', CS, ...
    'ColumnNames', {'phi1', 'Phi', 'phi2', 'weights'},'lsqnonneg');
```

**Comment by YoshieraHuang** [comment]
> Adding the option `'lsqnonneg'` should resolve your issue, i.e.,
> 
> ```matlab
> odf = loadODF_generic(fname, 'cs', CS, ...
>     'ColumnNames', {'phi1', 'Phi', 'phi2', 'weights'},'lsqnonneg');
> ```

Thanks for your reply! But the `lsqnonneg` option does not work. The results do not change.

**Comment by ralfHielscher** [comment]
Could you please post the exact output of the `loadODF` command.

Thanks,

Ralf.

PS: I have tried it now several times but I could not reproduce your result when using the option `lsqnonneg`

**Comment by YoshieraHuang** [comment]
> Could you please post the exact output of the `loadODF` command.
> 
> Thanks,
> 
> Ralf.
> 
> PS: I have tried it now several times but I could not reproduce your result when using the option `lsqnonneg`

I used wrong pictures for PFs and IPFs of restored ODF in the first comment of this issue. I edit it. The difference of ODs of three equivelant directions is not so large but is still not negligible.

In my last comment, I mistake the `lsqnonneg` option to `lsqnonneq` option. :sweat:. The output changes with `lsqnonneg` option but still not consitent with the input ODF:
![Cu recrystal odf txt lsg PF](https://user-images.githubusercontent.com/38752027/85939083-34ec3a00-b945-11ea-9b88-c2c73a2686ac.png)
![Cu recrystal odf txt lsg IPF](https://user-images.githubusercontent.com/38752027/85939079-31f14980-b945-11ea-9316-4b5aae413593.png)

Here is the output of the `loadODF` command without `lsqnonneg` option:
```
Interpolating the ODF. This might take some time...
   lsqr tolerance cut back: 0.013
   lsqr tolerance cut back: 0.017
   lsqr tolerance cut back: 0.022
 
odf = ODF (show methods, plot)
  crystal symmetry : Copper (m-3m)
  specimen symmetry: 1
 
  Uniform portion:
    weight: 0
 
  Radially symmetric portion:
    kernel: de la Vallee Poussin, halfwidth 5°
    center: 2432 orientations, resolution: 5°
    weight: 1
```

Here is the output with `lsqnonneg` option:
```
Interpolating the ODF. This might take some time...
 
odf = ODF (show methods, plot)
  crystal symmetry : Copper (m-3m)
  specimen symmetry: 1
 
  Uniform portion:
    weight: 0
 
  Radially symmetric portion:
    kernel: de la Vallee Poussin, halfwidth 5°
    center: 9 orientations, resolution: 5°
    weight: 1
```

The textureindex of input ODF, restored ODF without `lsqnonneg` and restored ODF with `lsqnonneg` are ~202, ~331 and ~166, respectively.

My version of MATLAB is R2020a.

**Comment by ralfHielscher** [comment, marked as answer]
It won't become better. Your initial ODF has a halfwidth of about 5 degree
and you export it on a  5 degree grid. This like you want to interpolate a
Gaussian from three points. You will need more points for higher accuracy.

Ralf


********************************************************************
Ralf Hielscher                   Tel: +371-531-38556
Fakultät für Mathematik               +371-531-22200 (Sekr.)
Technische Universität Chemnitz  Fax: +371-531-22109
Reichenhainer Str. 39            E-mail:
ralf.hielscher@mathematik.tu-chemnitz.de
D-09126 Chemnitz                 http://www.tu-chemnitz.de/~rahi
********************************************************************

**Comment by YoshieraHuang** [comment]
Thank you for your generous replies! I change the 5 degree grid to 1 degree grid, and here is the output:
![Cu recrystal odf 1 txt lsq PF](https://user-images.githubusercontent.com/38752027/85940157-53563380-b94d-11ea-8149-fabe3671c41b.png)
![Cu recrystal odf 1 txt lsq IPF](https://user-images.githubusercontent.com/38752027/85940158-53eeca00-b94d-11ea-8ef5-ac5b410bd99a.png)

The ODs are smaller than input ones, and the difference shoule be thought negligable? Could you please explain why the difference happens? Thank you

**Comment by ralfHielscher** [comment]
May I ask what you want to achieve? If you simply want to store an ODF
variable on disk and reload it later you can simply take the Matlab
commands save and load. This works without any error whatsoever.

The export command is thought to export ODFs to an external program. And
yes, this reconstruction is not perfect. But in your case it is quite ok.
The maximum value of an ODF, which you seem to compare, is the worst of all
ODF characteristics. As it is extremely sensitive to all kinds of
parameters choice. I would never ever use this value to characterize an ODF.

Ralf.



********************************************************************
Ralf Hielscher                   Tel: +371-531-38556
Fakultät für Mathematik               +371-531-22200 (Sekr.)
Technische Universität Chemnitz  Fax: +371-531-22109
Reichenhainer Str. 39            E-mail:
ralf.hielscher@mathematik.tu-chemnitz.de
D-09126 Chemnitz                 http://www.tu-chemnitz.de/~rahi
********************************************************************

**Comment by YoshieraHuang** [comment]
> May I ask what you want to achieve? If you simply want to store an ODF variable on disk and reload it later you can simply take the Matlab commands save and load. This works without any error whatsoever. The export command is thought to export ODFs to an external program. And yes, this reconstruction is not perfect. But in your case it is quite ok. The maximum value of an ODF, which you seem to compare, is the worst of all ODF characteristics. As it is extremely sensitive to all kinds of parameters choice. I would never ever use this value to characterize an ODF. Ralf. ******************************************************************** Ralf Hielscher Tel: +371-531-38556 Fakultät für Mathematik +371-531-22200 (Sekr.) Technische Universität Chemnitz Fax: +371-531-22109 Reichenhainer Str. 39 E-mail: ralf.hielscher@mathematik.tu-chemnitz.de D-09126 Chemnitz http://www.tu-chemnitz.de/~rahi ********************************************************************
> […](#)
> On Sun, Jun 28, 2020 at 8:45 AM Yoshiera ***@***.***> wrote: Thank you for your generous replies! I change the 5 degree grid to 1 degree grid, and here is the output: [image: Cu recrystal odf 1 txt lsq PF] <https://user-images.githubusercontent.com/38752027/85940157-53563380-b94d-11ea-8149-fabe3671c41b.png> [image: Cu recrystal odf 1 txt lsq IPF] <https://user-images.githubusercontent.com/38752027/85940158-53eeca00-b94d-11ea-8ef5-ac5b410bd99a.png> The ODs are smaller than input ones, and the difference shoule be thought negligable? Could you please explain why the difference happens? Thank you — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#493 (comment)](https://github.com/mtex-toolbox/mtex/issues/493#issuecomment-650706206)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACBYJ2K476LWIREFBOKWRTLRY3RHFANCNFSM4OKAOS2Q> .

I use MTEX to generate ODF for the input of my own program.
 
I compared the intensities of three equivalent crystal direction <100>, <010> and <001> in PF, which shoule be identical for cubic in my opinion. The arrows in this pictures show what i concerned.
![PF](https://user-images.githubusercontent.com/38752027/85942106-c9609780-b959-11ea-90b4-fc7547f7028c.png)

 The max intensities in IPFs are also related to this.
